prov/efa/src/rdm/rxr_pkt_pool.c

```diff
@@ -115,24 +184,144 @@ int rxr_pkt_pool_grow(struct rxr_pkt_pool *rxr_pkt_pool)
 {
 	int err;
 
-	err = ofi_bufpool_grow(rxr_pkt_pool->localinfo_pool);
+	err = ofi_bufpool_grow(rxr_pkt_pool->entry_pool);
 	if (err)
 		return err;
 
-	err = ofi_bufpool_grow(rxr_pkt_pool->wiredata_pool);
-	if (err)
-		return err;
+	if (rxr_pkt_pool->sendv_pool) {
+		err = ofi_bufpool_grow(rxr_pkt_pool->sendv_pool);
+		if (err)
+			return err;
+	}
 
 	return 0;
 }
 
+/**
+ * @brief post multiple packet entries to receive messages from shm
+ *
+ * @param[in]		pkt_pool	pkt pool, whose usage must be SHM RX packet pool
+ * @param[in]		nrecv		number of receiving buffers to post
+ * @returns
+ * On success, return 0.
+ * On failure, return negative libfabric error code.
+ */
+static inline
+int rxr_pkt_pool_bulk_post_recv_shm(struct rxr_pkt_pool *pkt_pool, int nrecv)
+{
+	int irecv, err;
+	struct rxr_pkt_entry *pkt_entry;
+
+	assert(pkt_pool->usage == RXR_PKT_FROM_SHM_RX_POOL);
+
+	for (irecv = 0; irecv < nrecv; ++irecv) {
+		pkt_entry = rxr_pkt_entry_alloc(pkt_pool->ep, pkt_pool, RXR_PKT_FROM_SHM_RX_POOL);
+		pkt_entry->x_entry = NULL;
+
+		err = fi_recv(pkt_pool->ep->shm_ep, pkt_entry->wiredata, pkt_pool->ep->mtu_size, NULL, FI_ADDR_UNSPEC, pkt_entry);
+		if (OFI_UNLIKELY(err)) {
+			rxr_pkt_entry_release_rx(pkt_pool->ep, pkt_entry);
+			EFA_WARN(FI_LOG_EP_CTRL,
+				"failed to post buf for shm  %d (%s)\n", -err,
+				fi_strerror(-err));
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief post multiple packet entries to receive messages from EFA device
+ *
+ * @param[in]		pkt_pool	pkt pool, whose usage must be EFA RX packet pool
+ * @param[in]		nrecv		number of receiving buffers to post
+ * @returns
+ * On success, return 0.
+ * On failure, return negative libfabric error code.
+ */
+static inline
+int rxr_pkt_pool_bulk_post_recv_efa(struct rxr_pkt_pool *pkt_pool, int nrecv)
+{
+	int irecv, err;
+	struct rxr_pkt_entry *pkt_entry;
+	struct ibv_recv_wr *prv_recv_wr, *cur_recv_wr, *bad_wr;
+
+	assert(pkt_pool->usage == RXR_PKT_FROM_EFA_RX_POOL);
+	assert(pkt_pool->efa_recv_wr_pool);
+
+	prv_recv_wr = NULL;
+	for (irecv = 0; irecv < nrecv; ++irecv) {
+		pkt_entry = rxr_pkt_entry_alloc(pkt_pool->ep, pkt_pool, RXR_PKT_FROM_EFA_RX_POOL);
+		if (pkt_entry == NULL) {
+			EFA_WARN(FI_LOG_CQ, "allocating EFA rx pakcet entry failed!");
+			return -FI_ENOBUFS;
+		}
+
+		pkt_entry->x_entry = NULL;
+
+		cur_recv_wr = &pkt_pool->efa_recv_wr_pool[irecv].wr;
```

Comment:
Actually, shouldn't this be `cur_recv_wr = &pkt_pool->efa_recv_wr_pool[ofi_buf_index(pkt_entry)].wr`?

My understanding is as follows: we bulk post recv packets when `rxr_ep_progress_internal` is called for the first time. Then, every time we release a RX packet, we increment `ep->efa_rx_pkts_to_post`. In the next call to `rxr_ep_progress_internal`, we post `ep->efa_rx_pkts_to_post` packets and set `ep->efa_rx_pkts_to_post = 0`.

So when we call `rxr_pkt_entry_alloc` on line 255 here, the returned memory can be from anywhere in the RX packet pool. It's not guaranteed that the first call returns a pointer to the first memory chunk.

Reply:
the `recv_wr`  memory does not have to be adjacent to the actual packet pool.

`recv_wr` is used to store information of the buffer to be posted, then passed to `ibv_post_recv`. It is not necessary that `recv_wr` be part of the `rxr_pkt_entry`

Reply:
You are right. I initially thought the memory of `recv_wr` can be immediately reused after call to `ibv_post_recv`, because this is how efa provider in rdma-core is implemented.

However, after checking the man page of rdma-core, it clearly states that buffers of `recv_wr` can not be reused util we get the completion.

https://github.com/linux-rdma/rdma-core/blob/6644617f085050cff4d583546007ffe96a1a1697/libibverbs/man/ibv_post_recv.3#L52

Reply:
So we will need to allocate recv-wr memory and release the memory when we get completion

Reply:
Thanks. I'm still not sure why this wasn't caught by CI.

rdma-core is just copying the buffer address we send in ibv_recv_wr ([link](https://github.com/linux-rdma/rdma-core/blob/c18417c56773fd311794ed44b7a2a31cd44c893c/providers/efa/verbs.c#L2226)). In line 277, you are setting the buffer address to a memory region that's potentially still in use.. https://github.com/ofiwg/libfabric/blob/6a2eb1713bf12a890b48543bf93447056e7c79cb/prov/efa/src/rdm/rxr_pkt_pool.c#L279

Reply:
`wiredata` is part of `rxr_pkt_entry` and is not released until `rxr_pkt_entry_release` is called.

`wiredata` is memory is not part of `recv_wr`.

`sg_list` is part of `recv_wr`, it is used in  [link](https://github.com/linux-rdma/rdma-core/blob/c18417c56773fd311794ed44b7a2a31cd44c893c/providers/efa/verbs.c#L2226) on the fly.

When `ibv_post_recv` is returned, `recv_wr` is not used (for rdma-core's EFA provider)

Reply:
@sunkuamzn  I was not able to implement that change without impacting performance. So I dropped the 3 patches that implement `efa_recv_wr_pool`. I will do that in follow up patches.

Reply:
@sunkuamzn  I added a commit to ensure `rxr_pkt_entry` is multiple of 64.